librz/bin/pdb/pdb.c

```diff
@@ -1,1433 +1,589 @@
-		STypeInfo *member_info;
-		rz_list_foreach (members, member_iter, member_info) {
-			switch (type_info->leaf_type) {
-			case eLF_STRUCTURE_19:
-			case eLF_CLASS_19:
-			case eLF_STRUCTURE:
-			case eLF_CLASS:
-			case eLF_UNION:
-				if (build_member_format(member_info, &format, &member_names) == -1) { // if failed
-					goto fail; // skip to the next one, we can't build format from this
-				}
-				break;
-			default:
-				rz_warn_if_reached();
-			}
-			rz_strbuf_append(&member_names, " ");
-		}
-		if (format.len == 0) { // if it has no members, it's useless for us then?
-			goto fail;
-		}
-		char *sanitized_name = rz_str_sanitize_sdb_key(name);
-		pdb->cb_printf("pf.%s %s %s\n", sanitized_name, rz_strbuf_get(&format), rz_strbuf_get(&member_names));
-
-		if (to_free_name) { // name can be generated or part of the PDB data
-			RZ_FREE(name);
-		}
-		RZ_FREE(sanitized_name);
-		rz_strbuf_fini(&format);
-		rz_strbuf_fini(&member_names);
-
-	fail: // if we can't print whole type correctly, don't print at all
-		if (to_free_name) {
-			RZ_FREE(name);
-		}
-		rz_strbuf_fini(&format);
-		rz_strbuf_fini(&member_names);
-	}
-}
-
-/**
- * \brief Prints out all the type information in regular,json or pf format
- * 
- * \param pdb PDB information
+ * \brief return the output text for types in PDB
+ * \param db RzTypeDB
+ * \param pdb PDB instance
  * \param mode printing mode
+ * \return string of pdb types
  */
-static void print_types(const RzPdb *pdb, PJ *pj, const int mode) {
-	RzList *plist = pdb->pdb_streams;
-	STpiStream *tpi_stream = rz_list_get_n(plist, ePDB_STREAM_TPI);
-
-	if (!tpi_stream) {
+RZ_API RZ_OWN char *rz_bin_pdb_types_as_string(RZ_NONNULL const RzTypeDB *db, RZ_NONNULL const RzPdb *pdb, const RzOutputMode mode) {
+	rz_return_val_if_fail(db && pdb, NULL);
+	TpiStream *stream = pdb->s_tpi;
+	if (!stream) {
 		eprintf("There is no tpi stream in current pdb\n");
-		return;
+		return NULL;
 	}
 	switch (mode) {
-	case 'd': print_types_regular(pdb, tpi_stream->types); return;
-	case 'j': print_types_json(pdb, pj, tpi_stream->types); return;
-	case 'r': print_types_format(pdb, tpi_stream->types); return;
+	case RZ_OUTPUT_MODE_STANDARD:
+		return pdb_type_as_string_regular(db, pdb, stream->print_type);
+	case RZ_OUTPUT_MODE_JSON:
+		return pdb_type_as_string_json(db, pdb, stream->print_type);
+	default:
+		return NULL;
 	}
 }
 
-static void print_gvars(RzPdb *pdb, ut64 img_base, PJ *pj, int format) {
-	SStreamParseFunc *omap = 0, *sctns = 0, *sctns_orig = 0, *gsym = 0, *tmp = 0;
-	SIMAGE_SECTION_HEADER *sctn_header = 0;
-	SGDATAStream *gsym_data_stream = 0;
-	SPEStream *pe_stream = 0;
-	SGlobal *gdata = 0;
+/**
+ * \brief return the output text for global symbols in PDB
+ * 
+ * \param pdb PDB instance
+ * \param img_base image base addr
+ * \param mode print mode
+ * \return string of pdb symbols
+ */
+RZ_API RZ_OWN char *rz_bin_pdb_gvars_as_string(RZ_NONNULL const RzPdb *pdb, const ut64 img_base, RzOutputMode mode) {
```

Comment:
Also this one should be placed somewhere under `librz/core`, maybe `librz/core/cpdb.c` or something like that.

Reply:
Why? It doesn't print, it returns a `char*` string.